### PR TITLE
Fix backup/restore with plain rewritable disk in case there are empty files in backup

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix backup/restore with plain rewritable disk in case there are empty files in backup.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 13dbd31608f9abca889ec6ab7376a0773f56b112 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72858/13dbd31608f9abca889ec6ab7376a0773f56b112/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/72858/13dbd31608f9abca889ec6ab7376a0773f56b112/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by qoega** [comment]
Fail looks related..

**Comment by jkartseva** [comment]
[test_backup_to_disk](https://d1k2gkhrlfqv31.cloudfront.net/clickhouse-test-reports-private/21164/13dbd31608f9abca889ec6ab7376a0773f56b112/integration_tests__asan__old_analyzer__[6_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel0_0.log) should be fixed, but merging is blocked by private.

**Comment by jkartseva** [comment]
Several `test_backup_to_disk` tests fail on master if run repeatedly with the `--count` option.
https://pastila.nl/?000b93b2/d0d8a25a64f1357a7eca37ab2ed385d0#Mfmn2zDT4iOFmNKvnoBU3A==

File: src/Disks/ObjectStorages/MetadataStorageFromPlainRewritableObjectStorage.h [hunk 1]
```diff
@@ -4,11 +4,31 @@
...
+class MetadataStorageFromPlainRewritableObjectStorageTransaction final : public MetadataStorageFromPlainObjectStorageTransaction
+{
+public:
+    explicit MetadataStorageFromPlainRewritableObjectStorageTransaction(
+        MetadataStorageFromPlainObjectStorage & metadata_storage_, ObjectStoragePtr object_storage_)
+        : DB::MetadataStorageFromPlainObjectStorageTransaction(metadata_storage_, object_storage_)
+    {
+    }
+
+    void createEmptyFile(const std::string & path) override
```
Comment (jkartseva) [inline src/Disks/ObjectStorages/MetadataStorageFromPlainRewritableObjectStorage.h:22]: The other update operations of the disk are revertable and backed by an [Operation](https://github.com/ClickHouse/clickhouse-private/blob/master/src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.h) type with defined `undo` rollback method.
I think it should be the same here:  `removeObjectIfExists` should be called if the transaction is being rolled back, but it's probably outside the scope of the fix – I can create a follow-up.

**Review by jkartseva — APPROVED** [review]
The test can be extracted from the `test_backup_restore` (e.g., as `test_backup_restore_system_tables`)  so the fix is not slowed down by the `test_backup_restore` [flakiness](https://pastila.nl/?000b93b2/d0d8a25a64f1357a7eca37ab2ed385d0#Mfmn2zDT4iOFmNKvnoBU3A==) which should be fixed separately.

File: src/Disks/ObjectStorages/MetadataStorageFromPlainRewritableObjectStorage.h (hunk 1 above)
Comment (kssenii) [reply src/Disks/ObjectStorages/MetadataStorageFromPlainRewritableObjectStorage.h:22]: Fixed in https://github.com/ClickHouse/ClickHouse/pull/72884.

**Pull request merged**